### PR TITLE
Correctly handle spaces in option-values

### PR DESCRIPTION
**Pull request by goerz**
This fixes a bug in the parsing of keys with "advanced" options. For example, the following would be a valid entry in an `authorized_keys` file:

```
command="sh -c \"mysqldump db1 -u fred1 -p\"",no-x11-forwarding ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDDN2n2kt99/aYpPbOZRJeGV62R2R1h3b8CMX8fpcGGXHJAxFWg/Sz8qXcbOeTRKdnBIWlUrkDi/7hWK
```

The values for the `command` option may contain spaces, and also escaped quotes. As stated in `man sshd 8`:

> The options (if present) consist of comma-separated option
> specifications. No spaces are permitted, except within double quotes.

and

> command="command"
>     Specifies that the command is executed whenever this key is used
>     for authentication. [...]
>     A quote may be included in the command by quoting it with a
>     backslash.

In the current version of `sshkeys`, reading the above key would result in an
Exception,

```
ValueError: Could not parse key, too many fields
```

The parser has been rewritten to account for the presence of spaces and escaped quotes. The behavior is tested in a new test, `test_pubkey_line_with_quote_options`.

In addition to the above fix, the version information of the package is now defined in a `__version__` module attribute, in accordance with PEP 0396 and PEP 0386. The `setup.py` script obtains the version by parsing `sshkeys/__init__.py`.

### CODE REVIEW

**Comment by goerz** [comment]
I would be happy to have this pull request accepted and released as version 0.5, if you have a chance. I'd want to use `sshkeys` in another project where keys with the command option are used extensively, so it would be great to have a new version with the bugfix installable from PyPI

**Comment by mbr** [comment]
Guess that's what I get for being too lazy to write a parser. Thank you!

I'll give this a check, but I don't see why it couldn't be merged if it is correct. The only issue I am having with it is the version-detection in setup.py, I normally use a different way of setting these version strings (but I forgot to add it into sshkeys). Would you want me to revert your changes, cherry pick parts from the patch (which would drop you as a commit author =() or just resubmit the pull request without the version-string patches?

File: sshkeys/__init__.py [hunk 1]
```diff
@@ -54,50 +56,93 @@ def readable_fingerprint(self):
...
-            if fields[1].startswith('ssh-') or fields[1].startswith('ecdsa'):
-                # 2nd field is type string, there's a comment preprepended
-                opt_str, type_str, data64 = fields
-            else:
-                type_str, data64, comment = fields
+            comment = None
+        elif len(fields) == 3:
+            type_str, data64, comment = fields
+        else:
+            raise ValueError("Too many fields in key: %s"
```
Comment (mbr) [inline sshkeys/__init__.py]: I believe this error message is wrong, since maxsplit=2, this could only be raised if `len(fields)` is <= 1?

**Comment by goerz** [comment]
I reverted the merge of the version-number change.

The maxsplit parameter is a bit confusing. Doing n splits results in n+1 items

~~Therefore, I believe the code is correct. I've added a comment in the code to clarify this.~~
EDIT: Sorry, I completely misunderstood your comment. You're completely right, of course!
This is now fixed, see below.

**Comment by goerz** [comment]
I actually just found a bug in my parser, where backslashes that were not an escape for a quote would be dropped. Added two commits with a fix of this bug, and another test.

File: sshkeys/__init__.py (hunk 1 above)
Comment (goerz) [inline sshkeys/__init__.py]: See below, `maxsplit=2` produces up to 3 fields

Comment (goerz) [inline sshkeys/__init__.py]: You're absolutely correct! Sorry, I completely misunderstood your comment at first. This is fixed now.

**Comment by goerz** [comment]
I fixed the bug with the wrong "too many fields" error message. Thanks for pointing this out, and sorry I didn't understand your comment at first. I also made sure that various malformed keys consistently produce a ValueError with an appropriate error message, with a new test.

**Comment by mbr** [comment]
I've merged what I'm willing to merge, that is everything except the version string part (which I'll fix myself) and the `.gitignore` (Stuff like that belongs into `~/.gitgignore` in my opinion, I use the per-project ones only for really project-specific stuff).

I've also managed to preserve the authorship credit by rebasing =). Give it a look and if you're satisfied, I'll have a go at cutting a new release.

**Comment by goerz** [comment]
Looks good!

**Comment by mbr** [comment]
0.5 is up on PyPI.

**Comment by goerz** [comment]
Awesome, thanks!

**Pull request closed without merging**